### PR TITLE
bootctl: unlink and cleanup functions

### PR DESCRIPTION
**Pull request by lnussel**

### CODE REVIEW

File: src/basic/chase-symlinks.h [hunk 1]
```diff
@@ -34,5 +34,6 @@ int chase_symlinks_and_opendir(const char *path, const char *root, ChaseSymlinks
 int chase_symlinks_and_stat(const char *path, const char *root, ChaseSymlinksFlags chase_flags, char **ret_path, struct stat *ret_stat, int *ret_fd);
 int chase_symlinks_and_access(const char *path, const char *root, ChaseSymlinksFlags chase_flags, int access_mode, char **ret_path, int *ret_fd);
 int chase_symlinks_and_fopen_unlocked(const char *path, const char *root, ChaseSymlinksFlags chase_flags, const char *open_flags, char **ret_path, FILE **ret_file);
+int chase_symlinks_and_unlink(const char *path, const char *root, ChaseSymlinksFlags chase_flags, int unlink_flags, char **ret_path);
```
Comment (bluca) [inline src/basic/chase-symlinks.h:37]: could you please add some unit tests for this in `src/test/test-fs-util.c`

**Comment by bluca** [comment]
also need to rebase away first commit

File: src/boot/bootctl-status.c [hunk 2]
```diff
@@ -503,6 +504,248 @@ int verb_status(int argc, char *argv[], void *userdata) {
...
+        if (n == 0) {
+                (void) hashmap_remove(known_files, fn);
+                free(k);
+        } else if (!k) {
+                _cleanup_free_ char *t = NULL;
+
+                t = strdup(fn);
+                if (!t)
+                        return -ENOMEM;
+                r = hashmap_put(known_files, t, INT_TO_PTR(n));
```
Comment (bluca) [inline src/boot/bootctl-status.c:532]: `hashmap_put_strdup()`

File: src/boot/bootctl-status.c [hunk 3]
```diff
@@ -503,6 +504,248 @@ int verb_status(int argc, char *argv[], void *userdata) {
...
+                        return i;
+
+        return -1;
+}
+
+static int unlink_entry(const BootConfig *config, const char *root, const char *id) {
+        _cleanup_(hashmap_free_free_keyp) Hashmap *known_files = NULL;
+        const BootEntry *e = NULL;
+        int r;
+
```
Comment (bluca) [inline src/boot/bootctl-status.c:644]: assert(config)

File: man/bootctl.xml [hunk 4]
```diff
@@ -390,6 +405,14 @@
         the firmware's boot option menu.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--dry</option></term>
```
Comment (bluca) [inline man/bootctl.xml]: bikeshedding: spell it out as `--dry-run`? that's how it's usually called

File: man/bootctl.xml [hunk 5]
```diff
@@ -153,6 +153,21 @@
         disables the timeout while always showing the menu. When an empty string ("") is specified the
         bootloader will revert to its default menu timeout.</para></listitem>
       </varlistentry>
+
+      <varlistentry>
+        <term><option>unlink</option> <replaceable>ID</replaceable></term>
```
Comment (bluca) [inline man/bootctl.xml:158]: please add all these new options/verbs to the shell completion scripts

File: src/boot/bootctl-status.c (hunk 2 above)
Comment (lnussel) [reply src/boot/bootctl-status.c:532]: that's something else. this hashmap has paths and the value is not a string to dup

File: man/bootctl.xml (hunk 5 above)
Comment (lnussel) [reply man/bootctl.xml:158]: *sigh* how many slices does the salami have?

Comment (bluca) [reply man/bootctl.xml:158]: hopefully enough to keep everyone well fed!

**Review by bluca — APPROVED** [review]

**Comment by bluca** [comment]
@mrc0mmand is the centos issue a known one? doesn't look related

**Comment by mrc0mmand** [comment]
> @mrc0mmand is the centos issue a known one? doesn't look related

No idea what happened there. I restarted the job, hopefully it was just a one-time hiccup.

**Pull request merged**